### PR TITLE
Add connection pool to support GraphClient thread safe and graphd failover able

### PR DESCRIPTION
**Pull request by chenxu14**
Some issus like #103 #88 #94 have already been resolved in our own branch.
Make a PR here, In order for anyone in need to reuse it.
And hope you guys can review it. @dutor @darionyaphet

### CODE REVIEW

File: client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java:1]: 2020

**Comment by darionyaphet** [comment]
Thank you for you take care of it :D

File: client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    private String user;
+    private String password;
+    private String space;
+    private int timeout;
+    private int connectionRetry;
+    private int executionRetry;
+    private int failoverRetry;
+    private int poolSize;
+
+    ConnectionContext withHosts(String hosts) {
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java:23]: `hosts` should be a list.

File: client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        this.hosts = hosts;
+        return this;
+    }
+
+    ConnectionContext withUser(String user) {
+        this.user = user;
+        return this;
+    }
+
+    ConnectionContext withPassword(String password) {
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java:33]: If the password is an empty string ... we should support more testing about the argument.

File: client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        this.timeout = Integer.parseInt(timeout);
+        return this;
+    }
+
+    ConnectionContext withConnectionRetry(String connectionRetry) {
+        this.connectionRetry = Integer.parseInt(connectionRetry);
+        return this;
+    }
+
+    ConnectionContext withExecutionRetry(String executionRetry) {
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java:48]: why don't support an integer as the argument?

File: client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+public class PooledGraphClient implements GraphClient {
+    private static final Logger LOG = LoggerFactory.getLogger(PooledGraphClient.class);
+    private final GenericObjectPool<GraphClientImpl> pool;
+    private final ConnectionContext context;
+    private volatile boolean closed = false;
+
+    public PooledGraphClient(Properties props) {
+        this.context = new ConnectionContext()
+                .withHosts(props.getProperty(GraphConstants.CONN_ADDRESS))
+                .withConnectionRetry(props.getProperty(GraphConstants.CONN_RETRY_COUNT, "4"))
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java:29]: So the connection pool's config is immutable ?

File: client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+import com.vesoft.nebula.utils.GraphConstants;
+import java.util.Properties;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Graph Client with thread safe and connection failover able
+ */
+public class PooledGraphClient implements GraphClient {
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java:20]: I don't know why the connection pool will implement Client

File: client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java (hunk 2 above)
Comment (chenxu14) [reply client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java:23]: hosts is named in the format “host:port,host2:port2”, and will parse to List in GraphClientFactory#parseHosts

File: client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java (hunk 4 above)
Comment (chenxu14) [reply client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java:48]: Would it be better to replace Property with org.apache.commons.configuration.Configuration? In this case, getInt can be used.

File: client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java (hunk 5 above)
Comment (chenxu14) [reply client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java:29]: currently, only hosts are updatable.

File: client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java (hunk 6 above)
Comment (chenxu14) [reply client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java:20]: In order to unify the upper interface

File: client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java (hunk 5 above)
Comment (darionyaphet) [reply client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java:29]: It's unnecessary...

File: client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java (hunk 6 above)
Comment (darionyaphet) [reply client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java:20]: It's unnecessary... we will using connection pool to get a client from it.

File: client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java (hunk 4 above)
Comment (darionyaphet) [reply client/src/main/java/com/vesoft/nebula/client/graph/ConnectionContext.java:48]: Sorry to reply late. I think Apache Configuration is good choice:D

File: client/src/main/java/com/vesoft/nebula/client/graph/NGQLException.java [hunk 7]
```diff
@@ -10,15 +10,20 @@
  *
  */
 public class NGQLException extends Exception {
+    private static final long serialVersionUID = -90563191494716743L;
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/client/graph/NGQLException.java:13]: `serialVersionUID` is used to serialization and deserialization. Don't add for each class.

**Comment by chenxu14** [comment]
OK，will be back soon.

**Comment by chenxu14** [comment]
Feature seems to be conflict with #122

**Comment by darionyaphet** [comment]
Would you still focus on this PR?

**Comment by chenxu14** [comment]
> Would you still focus on this PR?

If enable this we should revert #122, do you thing that's ok?

**Comment by darionyaphet** [comment]
> > Would you still focus on this PR?
> 
> If enable this we should revert #122, do you thing that's ok?

No they are not the same. `Collection Pool` major focus on reuse.

**Comment by darionyaphet** [comment]
refer [Add connection pool and LoadBalancer for java client](https://github.com/vesoft-inc/nebula-clients/pull/11)

**Pull request closed without merging**